### ISSUE TITLE
Exhaustion is prevented by tick

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/dataflower#104]
By the exhaustion example you got error, because the pull callback creates a deferred pull just before the flow gets exhausted.

**Comment by inf3rno** [comment]
```js
var ArrayReader = DataFlow.extend({
    prototype: {
        init: function (items) {
            DataFlow.prototype.init.call(this);
            if (!items.length) {
                this.exhaust();
                return;
            }
            var pointer = 0;
            this.on("pulled", function (flow) {
                setTimeout(function () {
                    flow.write(items[pointer]);
                    if (++pointer == items.length)
                        flow.exhaust();
                }, 1);
            });
        }
    }
});

var reader = new ArrayReader([1, 2, 3]);

reader.pull(function pump(item) {
    display.data(item);
    if (!reader.isExhausted())
        reader.pull(pump);
});
```

```js
        write: function (data) {
            if (this.isExhausted())
                throw new DataFlow.ExhaustedWrite(this);
            this.dataQueue.push(data);
            this.tick();
        },
        tick: function () {
            while (this.callbackQueue.length && this.dataQueue.length) {
                var callback = this.callbackQueue.shift();
                var data = this.dataQueue.shift();
                callback(data);
            }
            if (this.isDry())
                this.emit("dry", this);
        }
```

**Comment by inf3rno** [comment]
I am not sure what would be the proper solution here. I need to inject a write after exhaust, which according to the current rules is not possible. I don't want to loosen on the rules, so I need a workaround.

A possible solution to do the write without a tick and call the tick only after the exhaust. To do that I'll need a _write or something like that.

**Comment by inf3rno** [comment]
It is not possible to remove the tick from write unless if I tick manually.

**Comment by inf3rno** [comment]
I don't think it is possible to do any good with manual ticking:

```js
flow.await(console.log);
flow.write(1);
flow.write(2);
// nothing happens
flow.read();
flow.tick(); // 2
```

Ye ofc. read could trigger the tick, but what if I have a deferred tick when I need to call a write and an exhaust without a tick between the two? Problem.

I need this to be atomic, but adding a lastWrite method or something like that would not be wise, since I'd need to add a lastPush and so on... Adding a write flag like: `{exhausted: true}` is another possible option, but it would make code less readable.

**Comment by inf3rno** [comment]
What else can I do?
Let another pull maybe. :S Or just reject the last pull. I am not sure.

**Comment by inf3rno** [comment]
We could use a before tick callback by write:

```js
var ArrayReader = DataFlow.extend({
    prototype: {
        init: function (items) {
            DataFlow.prototype.init.call(this);
            if (!items.length) {
                this.exhaust();
                return;
            }
            var pointer = 0;
            this.on("pulled", function (flow) {
                setTimeout(function () {
                    flow.write(items[pointer], function (flow){
                        if (++pointer == items.length)
                            flow.exhaust();
                    });
                }, 1);
            });
        }
    }
});

var reader = new ArrayReader([1, 2, 3]);

reader.pull(function pump(item) {
    display.data(item);
    if (!reader.isExhausted())
        reader.pull(pump);
});
```

**Comment by inf3rno** [comment]
Actually I can live with that.

**Comment by inf3rno** [comment]
Another possible option:

```js
var ArrayReader = DataFlow.extend({
    prototype: {
        init: function (items) {
            DataFlow.prototype.init.call(this);
            if (!items.length) {
                this.exhaust();
                return;
            }
            var pointer = 0;
            this.on("pulled", function (flow) {
                setTimeout(function () {
                    flow.stopClock();
                    flow.write(items[pointer]);
                    if (++pointer == items.length)
                        flow.exhaust();
                    flow.tick();
                }, 1);
            });
        }
    }
});

var reader = new ArrayReader([1, 2, 3]);

reader.pull(function pump(item) {
    display.data(item);
    if (!reader.isExhausted())
        reader.pull(pump);
});
```

I don't like this, there is no guarantee that the clock will start again if somebody forgets tick() it will cause serious problems.


```js
var ArrayReader = DataFlow.extend({
    prototype: {
        init: function (items) {
            DataFlow.prototype.init.call(this);
            if (!items.length) {
                this.exhaust();
                return;
            }
            var pointer = 0;
            this.on("pulled", function (flow) {
                setTimeout(function () {
                    flow.commit(function (){
                        flow.write(items[pointer]);
                        if (++pointer == items.length)
                            flow.exhaust();
                    });
                }, 1);
            });
        }
    }
});

var reader = new ArrayReader([1, 2, 3]);

reader.pull(function pump(item) {
    display.data(item);
    if (!reader.isExhausted())
        reader.pull(pump);
});
```

**Comment by inf3rno** [comment]
What's the problem with this? There are event handlers. So if there is no tick, then these event handlers will run before tick. Still the write callback is the best solution.
